File: BPMOPMobile_VietNamAirLines/BPMOPMobile_VietNamAirLines/BPMOPMobile.Droid/Presenter/SharedView/SharedView_PopupControlTextInputFormat.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.Graphics;
using Android.OS;
using Android.Runtime;
using Android.Support.V7.App;
using Android.Views;
using Android.Widget;
using BPMOPMobile.Bean;
using BPMOPMobile.Droid.Class;
using BPMOPMobile.Droid.Core.Common;
using BPMOPMobile.Droid.Presenter.Fragment;
using Jp.Wasabeef;
using Newtonsoft.Json.Linq;

namespace BPMOPMobile.Droid.Presenter.SharedView
{
    class SharedView_PopupControlTextInputFormat : SharedView_PopupControlBase
    {
        public SharedView_PopupControlTextInputFormat(LayoutInflater _inflater, AppCompatActivity _mainAct, CustomBaseFragment _fragment, string _fragmentTag, View _rootView) : base(_inflater, _mainAct, _fragment, _fragmentTag, _rootView)
        {

        }

        public override void InitializeValue_Master(ViewElement clickedElement)
        {
            base.InitializeValue_Master(clickedElement);
        }

        public override void InitializeValue_InputGridDetail(ViewElement elementParent, ViewElement elementPopup, JObject JObjectChild)
        {
            base.InitializeValue_InputGridDetail(elementParent, elementPopup, JObjectChild);
        }

        public override void InitializeView()
        {
            base.InitializeView();
            try
            {
                bool isChanged = true;

                #region Get View - Init Data
                View _viewPopupControl = _inflater.Inflate(Resource.Layout.PopupControl_InputTextFormat, null);
                RelativeLayout _relaContent = _viewPopupControl.FindViewById<RelativeLayout>(Resource.Id.rela_PopupControl_InputTextFormat_Content);
                LinearLayout _lnContentClick = _viewPopupControl.FindViewById<LinearLayout>(Resource.Id.ln_PopupControl_InputTextFormat_ContentClick);
                ImageView _imgBack = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_Close);
                TextView _tvTitle = _viewPopupControl.FindViewById<TextView>(Resource.Id.tv_PopupControl_InputTextFormat_Title);
                ImageView _imgDone = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_Done);
                ImageView _imgDelete = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_Delete);
                RichEditor mEditor = _viewPopupControl.FindViewById<RichEditor>(Resource.Id.editor_PopupControl_InputTextFormat);
                HorizontalScrollView _scrollAction = _viewPopupControl.FindViewById<HorizontalScrollView>(Resource.Id.scroll_PopupControl_InputTextFormat_Action);

                _imgDelete.Visibility = ViewStates.Gone;
                _lnContentClick.Visibility = ViewStates.Gone;

                switch (base.flagView)
                {
                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailCreateTask:
                    default:
                        {
                            _tvTitle.Text = elementParent.Title;

                            if (!String.IsNullOrEmpty(elementParent.Value))
                            {
                                mEditor.SetHtml(elementParent.Value);
                            }

                            break;
                        }
                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                        {
                            _tvTitle.Text = elementPopup.Title;

                            if (!String.IsNullOrEmpty(elementPopup.Value))
                            {
                                mEditor.SetHtml(elementPopup.Value);
                            }

                            break;
                        }
                }
                #endregion

                #region Show View                
                Dialog _dialogPopupControl = new Dialog(_rootView.Context, Resource.Style.Theme_Custom_BPMOP_Dialog_FullScreen);
                Window window = _dialogPopupControl.Window;
                _dialogPopupControl.RequestWindowFeature(1);
                _dialogPopupControl.SetCanceledOnTouchOutside(false);
                _dialogPopupControl.SetCancelable(true);
                window.SetSoftInputMode(SoftInput.AdjustPan); // Để Hide soft keyboard ko bị Resize
                window.SetGravity(GravityFlags.Bottom);
                var dm = _mainAct.Resources.DisplayMetrics;

                _dialogPopupControl.SetContentView(_viewPopupControl);
                _dialogPopupControl.Show();
                WindowManagerLayoutParams s = window.Attributes;
                s.Width = WindowManagerLayoutParams.MatchParent;
                s.Height = WindowManagerLayoutParams.MatchParent;
                window.Attributes = s;
                #endregion

                #region Event

                #region Editor Library
                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_undo).Click += delegate { mEditor.Undo(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_redo).Click += delegate { mEditor.Redo(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_bold).Click += delegate { mEditor.SetBold(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_italic).Click += delegate { mEditor.SetItalic(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_subscript).Click += delegate { mEditor.SetSubscript(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_superscript).Click += delegate { mEditor.SetSuperscript(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_strikethrough).Click += delegate { mEditor.SetStrikeThrough(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_underline).Click += delegate { mEditor.SetUnderline(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_heading1).Click += delegate { mEditor.SetHeading(1); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_heading2).Click += delegate { mEditor.SetHeading(2); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_heading3).Click += delegate { mEditor.SetHeading(3); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_heading4).Click += delegate { mEditor.SetHeading(4); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_heading5).Click += delegate { mEditor.SetHeading(5); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_heading6).Click += delegate { mEditor.SetHeading(6); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_txt_color).Click += delegate { mEditor.SetTextColor(isChanged ? Color.Black : Color.Red); isChanged = !isChanged; };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_bg_color).Click += delegate { mEditor.SetTextBackgroundColor(isChanged ? Color.Transparent : Color.Yellow); isChanged = !isChanged; };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_indent).Click += delegate { mEditor.SetIndent(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_outdent).Click += delegate { mEditor.SetOutdent(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_align_left).Click += delegate { mEditor.SetAlignLeft(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_align_center).Click += delegate { mEditor.SetAlignCenter(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_align_right).Click += delegate { mEditor.SetAlignRight(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_blockquote).Click += delegate { mEditor.SetBlockquote(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_insert_bullets).Click += delegate { mEditor.SetBullets(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_insert_bullets).Click += delegate { mEditor.SetBullets(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_insert_numbers).Click += delegate { mEditor.SetNumbers(); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_insert_image).Click += delegate { mEditor.InsertImage("http://www.1honeywan.com/dachshund/image/7.21/7.21_3_thumb.JPG", "dachshund"); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_insert_link).Click += delegate { mEditor.InsertLink("https://github.com/wasabeef", "wasabeef"); };

                _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_InputTextFormat_insert_checkbox).Click += delegate { mEditor.InsertTodo(); };
                #endregion

                _imgBack.Click += (sender, e) =>
                {
                    CmmDroidFunction.HideSoftKeyBoard(_mainAct);
                    _dialogPopupControl.Dismiss();
                };

                _imgDelete.Click += (sender, e) =>
                {
                    mEditor.SetHtml("");
                    _imgDone.PerformClick();
                };

                _imgDone.Click += delegate
                {
                    CmmDroidFunction.HideSoftKeyBoard(_mainAct);
                    string _result = "";
                    if (!String.IsNullOrEmpty(mEditor.GetHtml().ToString()))
                    {
                        _result = mEditor.GetHtml().ToString();
                    }

                    switch (base.flagView)
                    {
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                            {
                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForElement(elementParent, _result);
                                _frag._adapterDetailExpandControl.NotifyDataSetChanged();
                                _dialogPopupControl.Dismiss();
                                break;
                            }
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                            {
                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForPopupGridDetail(elementParent, elementPopup, JObjectChild, _result);
                                _dialogPopupControl.Dismiss();
                                break;
                            }
                    }
                };
                #endregion

                CmmDroidFunction.ShowSoftKeyBoard(_rootView, _mainAct);
            }
            catch (Exception ex)
            {
#if DEBUG
                CmmDroidFunction.WriteTrackingError(this.GetType().Name, "InitializeView", ex);
#endif
            }
        }
    }
}

File: VNABPMOP/app/src/main/java/com/vuthao/bpmop/shareview/SharedView_PopupControlTextInputFormat.java
package com.vuthao.bpmop.shareview;

import android.app.Activity;
import android.app.Dialog;
import android.content.Intent;
import android.graphics.Color;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.google.gson.Gson;
import com.vuthao.bpmop.R;
import com.vuthao.bpmop.base.BroadcastUtility;
import com.vuthao.bpmop.base.Functions;
import com.vuthao.bpmop.base.custom.editor.RichEditor;
import com.vuthao.bpmop.base.keyboard.KeyboardManager;
import com.vuthao.bpmop.base.model.dynamic.ViewElement;
import com.vuthao.bpmop.base.vars.VarsReceiver;
import com.vuthao.bpmop.core.Vars;

import org.json.JSONObject;

public class SharedView_PopupControlTextInputFormat extends SharedView_PopupControlBase implements View.OnClickListener {
    private boolean isChange;
    private String result;

    private Dialog dialog;
    private View view;
    private LinearLayout lnContentClick;
    private ImageView imgBack;
    private TextView tvTitle;
    private ImageView imgDone;
    private ImageView imgDelete;
    private RichEditor mEditor;

    public SharedView_PopupControlTextInputFormat(LayoutInflater inflater, Activity mainAct, String fragmentTag, View rootView) {
        super(inflater, mainAct, fragmentTag, rootView);
    }

    @Override
    public void initializeValue_Master(ViewElement viewElement) {
        super.initializeValue_Master(viewElement);
    }

    @Override
    public void initializeValue_InputGridDetail(ViewElement elementParent, ViewElement elementPopup, JSONObject JObjectChild) {
        super.initializeValue_InputGridDetail(elementParent, elementPopup, JObjectChild);
    }

    @Override
    public void initializeView() {
        super.initializeView();

        view = inflater.inflate(R.layout.popup_control_text_input_format, null);
        lnContentClick = view.findViewById(R.id.ln_PopupControl_InputTextFormat_ContentClick);
        imgBack = view.findViewById(R.id.img_PopupControl_InputTextFormat_Close);
        tvTitle = view.findViewById(R.id.tv_PopupControl_InputTextFormat_Title);
        imgDone = view.findViewById(R.id.img_PopupControl_InputTextFormat_Done);
        imgDelete = view.findViewById(R.id.img_PopupControl_InputTextFormat_Delete);
        mEditor = view.findViewById(R.id.editor_PopupControl_InputTextFormat);

        ImageView undo = view.findViewById(R.id.img_PopupControl_InputTextFormat_undo);
        ImageView redo = view.findViewById(R.id.img_PopupControl_InputTextFormat_redo);
        ImageView bold = view.findViewById(R.id.img_PopupControl_InputTextFormat_bold);
        ImageView italic = view.findViewById(R.id.img_PopupControl_InputTextFormat_italic);
        ImageView subscript = view.findViewById(R.id.img_PopupControl_InputTextFormat_subscript);
        ImageView superscript = view.findViewById(R.id.img_PopupControl_InputTextFormat_superscript);
        ImageView strikethrough = view.findViewById(R.id.img_PopupControl_InputTextFormat_strikethrough);
        ImageView underline = view.findViewById(R.id.img_PopupControl_InputTextFormat_underline);
        ImageView heading1 = view.findViewById(R.id.img_PopupControl_InputTextFormat_heading1);
        ImageView heading2 = view.findViewById(R.id.img_PopupControl_InputTextFormat_heading2);
        ImageView heading3 = view.findViewById(R.id.img_PopupControl_InputTextFormat_heading3);
        ImageView heading4 = view.findViewById(R.id.img_PopupControl_InputTextFormat_heading4);
        ImageView heading5 = view.findViewById(R.id.img_PopupControl_InputTextFormat_heading5);
        ImageView heading6 = view.findViewById(R.id.img_PopupControl_InputTextFormat_heading6);
        ImageView txt_color = view.findViewById(R.id.img_PopupControl_InputTextFormat_txt_color);
        ImageView bg_color = view.findViewById(R.id.img_PopupControl_InputTextFormat_bg_color);
        ImageView indent = view.findViewById(R.id.img_PopupControl_InputTextFormat_indent);
        ImageView outdent = view.findViewById(R.id.img_PopupControl_InputTextFormat_outdent);
        ImageView left = view.findViewById(R.id.img_PopupControl_InputTextFormat_align_left);
        ImageView center = view.findViewById(R.id.img_PopupControl_InputTextFormat_align_center);
        ImageView right = view.findViewById(R.id.img_PopupControl_InputTextFormat_align_right);
        ImageView blockquote = view.findViewById(R.id.img_PopupControl_InputTextFormat_blockquote);
        ImageView bullets = view.findViewById(R.id.img_PopupControl_InputTextFormat_insert_bullets);
        ImageView numbers = view.findViewById(R.id.img_PopupControl_InputTextFormat_insert_numbers);
        ImageView image = view.findViewById(R.id.img_PopupControl_InputTextFormat_insert_image);
        ImageView link = view.findViewById(R.id.img_PopupControl_InputTextFormat_insert_link);
        ImageView checkbox = view.findViewById(R.id.img_PopupControl_InputTextFormat_insert_checkbox);

        imgDelete.setVisibility(View.GONE);
        lnContentClick.setVisibility(View.GONE);

        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow:
            default: {
                tvTitle.setText(elementParent.getTitle());
                if (!Functions.isNullOrEmpty(elementParent.getValue())) {
                    mEditor.setHtml(elementParent.getValue());
                }
                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                tvTitle.setText(elementPopup.getTitle());
                if (!Functions.isNullOrEmpty(elementPopup.getValue())) {
                    mEditor.setHtml(elementPopup.getValue());
                }
                break;
            }
        }

        // region Dialog
        dialog = new Dialog(mainAct, R.style.Theme_Custom_BPMOP_Dialog_FullScreen);
        Window window = dialog.getWindow();
        dialog.requestWindowFeature(1);
        dialog.setCanceledOnTouchOutside(false);
        dialog.setCancelable(true);
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN);
        window.setGravity(Gravity.BOTTOM);
        dialog.setContentView(view);
        dialog.show();

        WindowManager.LayoutParams s = window.getAttributes();
        s.width = WindowManager.LayoutParams.MATCH_PARENT;
        s.height = WindowManager.LayoutParams.MATCH_PARENT;
        window.setAttributes(s);
        // endregion

        // region Editor
        undo.setOnClickListener(v -> mEditor.undo());
        underline.setOnClickListener(v -> mEditor.setUnderline());
        redo.setOnClickListener(v -> mEditor.redo());
        bold.setOnClickListener(v -> mEditor.setBold());
        italic.setOnClickListener(v -> mEditor.setItalic());
        subscript.setOnClickListener(v -> mEditor.setSubscript());
        strikethrough.setOnClickListener(v -> mEditor.setStrikeThrough());
        superscript.setOnClickListener(v -> mEditor.setSuperscript());
        heading1.setOnClickListener(v -> mEditor.setHeading(1));
        heading2.setOnClickListener(v -> mEditor.setHeading(2));
        heading3.setOnClickListener(v -> mEditor.setHeading(3));
        heading4.setOnClickListener(v -> mEditor.setHeading(4));
        heading5.setOnClickListener(v -> mEditor.setHeading(5));
        heading6.setOnClickListener(v -> mEditor.setHeading(6));
        txt_color.setOnClickListener(v -> {
            mEditor.setTextColor(isChange ? Color.BLACK : Color.RED);
            isChange = !isChange;
        });
        bg_color.setOnClickListener(v -> {
            mEditor.setBackgroundColor(isChange ? Color.TRANSPARENT : Color.YELLOW);
            isChange = !isChange;
        });
        indent.setOnClickListener(v -> mEditor.setIndent());
        outdent.setOnClickListener(v -> mEditor.setOutdent());
        left.setOnClickListener(v -> mEditor.setAlignLeft());
        center.setOnClickListener(v -> mEditor.setAlignCenter());
        right.setOnClickListener(v -> mEditor.setAlignRight());
        blockquote.setOnClickListener(v -> mEditor.setBlockquote());
        bullets.setOnClickListener(v -> mEditor.setBullets());
        numbers.setOnClickListener(v -> mEditor.setNumbers());
        image.setOnClickListener(v -> mEditor.insertImage("http://www.1honeywan.com/dachshund/image/7.21/7.21_3_thumb.JPG", "dachshund"));
        link.setOnClickListener(v -> mEditor.insertLink("https://github.com/wasabeef", "wasabeef"));
        checkbox.setOnClickListener(v -> mEditor.insertTodo());

        //endregion

        imgBack.setOnClickListener(this);
        imgDelete.setOnClickListener(this);
        imgDone.setOnClickListener(this);

        KeyboardManager.showKeyBoard(mEditor, mainAct);
    }

    private void done() {
        KeyboardManager.hideKeyboard(mainAct);
        if (!Functions.isNullOrEmpty(mEditor.getHtml())) {
            result = mEditor.getHtml();
        }

        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow: {
                Intent intent = new Intent();
                intent.setAction(VarsReceiver.UPDATEFORM);
                intent.putExtra("element", new Gson().toJson(elementParent));
                intent.putExtra("newValue", result);
                BroadcastUtility.send(mainAct, intent);
                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                Intent intent = new Intent();
                intent.setAction(VarsReceiver.UPDATECHILDACTION);
                intent.putExtra("elementParent", new Gson().toJson(elementParent));
                intent.putExtra("elementChild", new Gson().toJson(elementPopup));
                intent.putExtra("json", JObjectChild.toString());
                intent.putExtra("newValue", result);
                BroadcastUtility.send(mainAct, intent);
                break;
            }
        }

        dialog.dismiss();
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.img_PopupControl_InputTextFormat_Close: {
                KeyboardManager.hideKeyboard(mEditor, mainAct);
                dialog.dismiss();
                break;
            }
            case R.id.img_PopupControl_InputTextFormat_Delete: {
                mEditor.setHtml("");
                break;
            }
            case R.id.img_PopupControl_InputTextFormat_Done: {
                done();
                break;
            }
        }
    }
}
